core/shared/src/test/scala/fs2/ChunkSpec.scala

```diff
@@ -179,4 +179,26 @@ class ChunkSpec extends Fs2Spec {
     "longs" in testEmptyConcat(Chunk.concatLongs)
     "chars" in testEmptyConcat(Chunk.concatChars)
   }
+
+  "map andThen toArray" in {
+    val arr: Array[Int] = Chunk(0, 0).map(identity).toArray
```

Comment:
I think it's worth adding test similar to the actual issue, i.e. `Chunk(0, 0).map(_.toByte).toArray`, since I think the issue is that it' inferred as `Int` initially, but then we're basically casting it into a byte, so the types change. I don't think identity will do anything like that in this case.

Reply:
The `map(_.toByte)` was a red herring -- the issue reproduces with `map(identity)` as well. I added another test that more directly captures the issue:

```
// This one works fine b/c the underlying `Array[Any]` is returned
Chunk.Boxed(Array[Any](0)).asInstanceOf[Chunk[Int]].toArray[Any]
// This one throws b/c the underlying `Array[Any]` is returned and is stored in a reference to an `Array[Int]`, which causes the `ClassCastException`
Chunk.Boxed(Array[Any](0)).asInstanceOf[Chunk[Int]].toArray[Int]
```